src/main/scala/com/fulcrumgenomics/bam/FindSwitchbackReads.scala

```diff
@@ -0,0 +1,395 @@
+    * from the start of the target sequence, or if `preferLaterMatches` is true, from the end of the target sequence.
+    *
+    * @param query the query sequence to be found
+    * @param target the target sequence in which the query sequence is to be located
+    * @param maxErrorRate the maximum rate of mismatches allowed when matching the sequences
+    * @param preferLaterMatches if true prefer matches closer to the end of targetSequence, else prefer matches closer
+    *                           to the start
+    * @return a [[ReadBasedHit]] if a match is found, otherwise None.
+    */
+  def find(query: String, target: String, maxErrorRate: Double, preferLaterMatches: Boolean = false): Option[ReadBasedHit] = {
+    val maxMismatches = Math.floor(query.length * maxErrorRate)
+    val range = {
+      if (preferLaterMatches) Range.inclusive(target.length-query.length, 0, step = -1)
+      else Range.inclusive(0, target.length-query.length-1, step=1)
+    }
+
+    range.iterator.flatMap { targetOffset =>
+      var mismatches = 0
+      forloop (from=0, until=query.length) { i =>
+        if (!SequenceUtil.basesEqual(query(i).toByte, target(targetOffset+i).toByte)) mismatches += 1
+      }
+
+      if (mismatches <= maxMismatches) Some(ReadBasedHit(offset=targetOffset, length=query.length))
+      else None
+    }.find(_ => true)
+  }
+
+  /** Returns the read number, either 1 or 2. */
+  private def readNum(rec: SamRecord): Int = if (!rec.paired || rec.firstOfPair) 1 else 2
+
+  /** Returns the concatenation of the unique values per read group, or the default if there are no values. */
+  private def summarizeRg(header: SAMFileHeader, f: SAMReadGroupRecord => String, default: String = "n/a"): String = {
+    val ss = header.getReadGroups.map(f).filter(_ != null).map(_.trim).filter(_.nonEmpty).toSeq.distinct.sorted
+    if (ss.nonEmpty) ss.mkString(",") else default
+  }
+}
+
+@clp(group=ClpGroups.Personal, description=
+  """
+    |Finds reads where a template switch occurred during library construction.  Some library construction methods,
+    |notably ultra-low-input shotgun methods, are prone to template switching events that create molecules
+    |(templates, inserts) that instead of being a linear copy of a segment of genomic DNA, instead are chimeras formed
+    |by starting on one strand of DNA and then, switching to the opposite strand.  Frequently when the switch occurs
+    |there may be a small offset between where the first strand was departed and the opposite strand joined.
+    |
+    |## Algorithm
+    |
+    |Templates that contain strand switch events (switch-backs) are found by this tool in two different ways:
+    |
+    |1. By looking at reads that contain soft-clipping at their 5' end that matches the genome, after
+    |   being reverse complemented, proximal to the 5'-most mapped base of the read.  We call these matches
```

Comment:
huh? so do you look at the 5' end of the read before or after reverse complementing.

Reply:
Do you mean:
```
By looking at reads that contain soft-clipping at their 5' end that matches the genome when reverse complemented.
```